test/e2e/network/loadbalancer.go

```diff
@@ -1218,172 +1218,155 @@ var _ = common.SIGDescribe("LoadBalancers ExternalTrafficPolicy: Local", feature
-		noEndpointNodeMap := make(map[string]string)
-		for _, node := range epNodes {
-			ips := e2enode.GetAddresses(&node, v1.NodeInternalIP)
-			if len(ips) < 1 {
-				framework.Failf("No internal ip found for node %s", node.Name)
-			}
-			endpointNodeMap[node.Name] = ips[0]
-		}
-		for _, n := range nodes.Items {
-			ips := e2enode.GetAddresses(&n, v1.NodeInternalIP)
-			if len(ips) < 1 {
-				framework.Failf("No internal ip found for node %s", n.Name)
-			}
-			if _, ok := endpointNodeMap[n.Name]; !ok {
-				noEndpointNodeMap[n.Name] = ips[0]
-			}
+		framework.ExpectNoError(err, "creating the service")
+		ingress := e2eservice.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
+		svcPort := int(svc.Spec.Ports[0].Port)
+		nodePortStr := fmt.Sprintf("%d", svc.Spec.Ports[0].NodePort)
+		hcNodePortStr := fmt.Sprintf("%d", svc.Spec.HealthCheckNodePort)
+		framework.Logf("ingress is %s:%d, NodePort is %s, HealthCheckNodePort is %s", ingress, svcPort, nodePortStr, hcNodePortStr)
+
+		// Get the (single) endpoint's node
+		var endpointNodeName string
+		endpointsNodeMap, err := getEndpointNodesWithInternalIP(ctx, jig)
+		framework.ExpectNoError(err, "fetching endpoint node IPs")
+		for node, nodeIP := range endpointsNodeMap {
+			framework.Logf("endpoint is on node %s (%s)", node, nodeIP)
+			endpointNodeName = node
+			break
 		}
-		gomega.Expect(endpointNodeMap).ToNot(gomega.BeEmpty())
-		gomega.Expect(noEndpointNodeMap).ToNot(gomega.BeEmpty())
 
-		svcTCPPort := int(svc.Spec.Ports[0].Port)
-		svcNodePort := int(svc.Spec.Ports[0].NodePort)
-		ingressIP := e2eservice.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
-		path := "/clientip"
-		dialCmd := "clientip"
+		// The execPod has to be on a different node from the endpoint so we don't
+		// hit the "short circuit" rule for local connections.
+		ginkgo.By("creating a HostNetwork exec pod on a different node")
+		execPod := launchHostExecPod(ctx, cs, namespace, "hostexec", &endpointNodeName)
+		hostExecPodNodeIP := execPod.Status.HostIP
+
+		checkAllNodes := func(svc *v1.Service) {
+			// We don't want to allow KubeProxyEndpointLagTimeout for *each*
+			// node; we only allow that interval total from the start of each
+			// checkAllNodes() call.
+			timeout := time.Now().Add(e2eservice.KubeProxyEndpointLagTimeout)
+
+			for i := range nodes.Items {
+				node := &nodes.Items[i]
+				ips := e2enode.GetAddresses(node, v1.NodeInternalIP)
+				if len(ips) < 1 {
+					framework.Failf("No internal ip found for node %s", node.Name)
+				}
 
-		config := e2enetwork.NewNetworkingTestConfig(ctx, f)
+				local := svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyLocal
+				hasEndpoint := node.Name == endpointNodeName
+
+				// We should be able to reach the service via this node's
+				// nodeport if (a) the service has Cluster traffic policy,
+				// (b) the service has an endpoint on this node, or (c)
+				// this node is the node that our execpod is on so we'll
```

Comment:
I think some network plugins like cilium don't implement the short-circuit rule, e2e tests define the conformance with SIG network defined semantics, should we avoid testing the short-circuit ? is this something we are going to require to network plugins?

Reply:
Ah, [the service proxy doc](https://github.com/kubernetes-sigs/kpng/blob/master/doc/service-proxy.md) mentions short-circuiting for LB IPs and ExternalIPs but not NodePorts. IIRC I argued that short-circuit for NodePorts was illogical but Tim argued that it was convenient for localhost NodePorts at least.

OK, let's treat it as undefined and just be careful to avoid testing that case.

Reply:
🤷